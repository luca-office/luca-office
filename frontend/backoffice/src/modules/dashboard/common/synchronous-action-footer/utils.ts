import {IconName, TimeUnit} from "shared/enums"
import {ProjectModule} from "shared/models"
import {secondsToGivenTimeUnit} from "shared/utils"

export const getSecondaryButtonIcon = (
  hasSurveyStarted: boolean,
  hasSurveyEnded: boolean,
  hasNextModule: boolean,
  isAsyncSurvey: boolean
) => {
  if (hasSurveyStarted && !hasSurveyEnded) {
    return hasNextModule && !isAsyncSurvey ? IconName.Play : IconName.Check
  } else if (hasSurveyEnded) {
    return undefined
  } else {
    return IconName.Play
  }
}

export const getModuleDurationInMinutes = ({scenario}: ProjectModule) => {
  if (scenario && scenario.maxDurationInSeconds !== null) {
    return secondsToGivenTimeUnit(TimeUnit.Minute, scenario.maxDurationInSeconds)
  } else {
    return null
  }
}
