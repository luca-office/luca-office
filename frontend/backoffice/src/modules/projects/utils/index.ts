export * from "./sorting"
export * from "./state"
