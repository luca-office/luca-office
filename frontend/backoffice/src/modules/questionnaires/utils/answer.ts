export const getScore = (scoreValue: string, lowerLimit = 0, upperLimit = 99): number => {
  const score = parseInt(scoreValue, 10) || lowerLimit
  return score < lowerLimit ? lowerLimit : score > upperLimit ? upperLimit : score
}
