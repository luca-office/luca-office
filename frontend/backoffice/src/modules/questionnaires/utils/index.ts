export * from "./answer"
export * from "./questions"
