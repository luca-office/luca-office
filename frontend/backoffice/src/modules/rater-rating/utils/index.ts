export * from "./coding-dimension"
export * from "./participant"
export * from "./progress-indicator-status"
export * from "./questions"
export * from "./rating-status"
