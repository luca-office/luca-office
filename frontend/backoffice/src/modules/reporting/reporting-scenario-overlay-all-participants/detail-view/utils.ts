import {sum} from "lodash"
import {TableEntity} from "shared/components/rating/models"
import {CodingDimension, CodingItem, CodingItemResultsByItemId} from "shared/models"

export const itemToTableEntity = (
  item: CodingItem,
  codingItemResultsByItem: CodingItemResultsByItemId
): TableEntity => ({
  id: item.id,
  title: item.title,
  position: item.position,
  score: codingItemResultsByItem[item.id]?.[0]?.score,
  maxScore: codingItemResultsByItem[item.id]?.[0]?.maximumScore,
  averageScore: codingItemResultsByItem[item.id]?.[0]?.averageScore,
  rated: true
})

export const dimensionToTableEntity = (
  dimension: CodingDimension,
  items: CodingItem[],
  codingItemResultsByItem: CodingItemResultsByItemId
): TableEntity => ({
  id: dimension.id,
  title: dimension.title,
  position: dimension.position,
  score: items.reduce(
    (acc, item) => acc + sum((codingItemResultsByItem[item.id] ?? []).map(result => result.score)),
    0
  ),
  maxScore: items.reduce((acc, item) => acc + codingItemResultsByItem[item.id]?.[0]?.maximumScore, 0),
  averageScore: items.reduce((acc, item) => acc + codingItemResultsByItem[item.id]?.[0]?.averageScore, 0),
  rated: true
})
