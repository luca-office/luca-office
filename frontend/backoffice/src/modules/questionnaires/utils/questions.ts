import {getQuestionTypeIconName} from "shared/components"
import {QuestionScoringType, QuestionType} from "shared/graphql/generated/globalTypes"
import {FreetextQuestionCodingCriterion, QuestionnaireAnswer, QuestionnaireQuestion} from "shared/models"
import {exists, Option} from "shared/utils"
import {ResortableEntity} from "../../../models"

export const convertQuestionsToResortableEntity = (
  questionsOption: Option<QuestionnaireQuestion[]>
): Option<ResortableEntity[]> =>
  questionsOption.map<ResortableEntity[]>(questions =>
    questions.map(question => ({
      id: question.id,
      position: question.position,
      icon: getQuestionTypeIconName(question.questionType),
      title: question.text
    }))
  )

export const convertAnswersToResortableEntity = (
  answersOption: Option<QuestionnaireAnswer[]>
): Option<ResortableEntity[]> =>
  answersOption.map<ResortableEntity[]>(answers =>
    answers.map(answer => ({
      id: answer.id,
      position: answer.position,
      title: answer.text
    }))
  )

/**
 * if the user changes the question type from a type that support a certain scoring to
 * a type that does not support scoring, we need to reset the scoring
 * @param isScoringEnabled - does the current question have a scoring setting !== None
 * @param sourceQuestionType - type of current question
 * @param targetQuestionType - selected type that should be switched to
 */
export const isScoringResetNecessary = (
  isScoringEnabled: boolean,
  sourceQuestionType: QuestionType,
  targetQuestionType: QuestionType
) => {
  return targetQuestionType !== QuestionType.FreeText && sourceQuestionType !== targetQuestionType && isScoringEnabled
}

const hasEmptyQuestionTitle = (questions: QuestionnaireQuestion[]) =>
  questions.some(question => question.text.trim() === "")

const hasEmptyQuestionAnswers = (questions: QuestionnaireQuestion[]) =>
  questions.some(question => question.answers.some(answer => answer.text.trim() === ""))

const isNoScoreFreetextCriterion = (criterion: FreetextQuestionCodingCriterion) =>
  criterion.description.trim() === "" && criterion.score === 0

const hasEmptyCodingCriteria = (questions: QuestionnaireQuestion[]) =>
  questions
    .filter(question => question.questionType === QuestionType.FreeText)
    .some(freeTextQuestion => {
      // only allow extact one criterion with empty description and score of 0
      return freeTextQuestion.freetextQuestionCodingCriteria.filter(isNoScoreFreetextCriterion).length > 1
        ? true
        : freeTextQuestion.freetextQuestionCodingCriteria
            .filter(criterion => !isNoScoreFreetextCriterion(criterion))
            .some(criterion => criterion.description.trim() === "") &&
            freeTextQuestion.scoringType !== QuestionScoringType.None
    })

export const containsEmptyTitleAnswersOrCriteria = (questions: QuestionnaireQuestion[]) =>
  hasEmptyQuestionTitle(questions) || hasEmptyQuestionAnswers(questions) || hasEmptyCodingCriteria(questions)

export const containsUnsetCorrectAnswer = (questions: QuestionnaireQuestion[]): boolean =>
  exists(question => {
    const needsCorrectAnswer =
      (question.questionType === QuestionType.SingleChoice || question.questionType === QuestionType.MultipleChoice) &&
      question.scoringType !== QuestionScoringType.None

    return needsCorrectAnswer && !exists(answer => answer.isCorrect, question.answers)
  }, questions)
