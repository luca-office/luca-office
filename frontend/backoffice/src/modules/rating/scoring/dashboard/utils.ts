import {isFuture, isToday} from "date-fns"
import {LucaTFunction} from "shared/translations"
import {formatDate} from "shared/utils"

export const getEndDateLabel = (date: Date, t: LucaTFunction) => {
  return isFuture(date)
    ? `${t("rating_overview__rating_end", {date: formatDate(date)})}`
    : `${t("rating_overview__rating_ended", {date: formatDate(date)})} ${
        isToday(date) ? `(${t("rating_overview__ended_today")})` : ""
      }`
}
