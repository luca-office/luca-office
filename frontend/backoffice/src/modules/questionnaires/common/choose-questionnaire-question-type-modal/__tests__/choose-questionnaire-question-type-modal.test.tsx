import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {Modal, Overlay, SelectableCard} from "shared/components"
import {Option} from "shared/utils"
import {
  ChooseQuestionnaireQuestionTypeModal,
  ChooseQuestionnaireQuestionTypeModalProps
} from "../choose-questionnaire-question-type-modal"

const defaultProps: ChooseQuestionnaireQuestionTypeModalProps = {
  disabled: false,
  onDismiss: jest.fn(),
  onConfirm: jest.fn(),
  selectedQuestionType: Option.none(),
  setSelectedQuestionType: jest.fn()
}

const getComponent = (props?: Partial<ChooseQuestionnaireQuestionTypeModalProps>) => (
  <ChooseQuestionnaireQuestionTypeModal {...{...defaultProps, ...props}} />
)

describe("choose-questionnaire-question-type-modal", () => {
  it("renders correctly", () => {
    const component = create(getComponent())

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure (no error message)", () => {
    const component = shallow(getComponent())

    const overlay = component.find(Overlay)
    expect(overlay).toHaveLength(1)

    const modal = overlay.dive().find(Modal)
    expect(modal).toHaveLength(1)

    const modalContent = modal.dive()
    const selectableCard = modalContent.find(SelectableCard)
    expect(selectableCard).toHaveLength(3)

    const errorMessage = modalContent.find(".error-message")
    expect(errorMessage).toHaveLength(0)
  })
  it("has correct structure (error message)", () => {
    const component = shallow(getComponent({errorMessage: "error-message"}))

    const overlay = component.find(Overlay)
    expect(overlay).toHaveLength(1)

    const modal = overlay.dive().find(Modal)
    expect(modal).toHaveLength(1)

    const modalContent = modal.dive()
    const selectableCard = modalContent.find(SelectableCard)
    expect(selectableCard).toHaveLength(3)

    const errorMessage = modalContent.find(".error-message")
    expect(errorMessage).toHaveLength(1)
  })
  it("triggers dismiss correctly", () => {
    const onDismissMock = jest.fn()
    const component = shallow(getComponent({onDismiss: onDismissMock}))

    const modal = component.find(Overlay).dive().find(Modal)
    expect(modal).toHaveLength(1)

    const onDismissHandler = modal.prop("onDismiss")
    expect(onDismissHandler).toBeDefined()

    onDismissHandler!()
    expect(onDismissMock).toHaveBeenCalledTimes(1)
  })
})
