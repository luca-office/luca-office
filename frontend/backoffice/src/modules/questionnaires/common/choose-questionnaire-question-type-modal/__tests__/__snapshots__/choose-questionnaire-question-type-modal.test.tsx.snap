// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choose-questionnaire-question-type-modal renders correctly 1`] = `
<div
  className="css-1iw6hgd-Overlay"
>
  <div
    className="css-1iw6hgd-Overlay"
  >
    <div>
      <div
        className="css-1cw1kcy-Modal"
      >
        <header
          className="css-14sl779-Modal"
        >
          <div
            className="css-1f6p2vj-Heading"
          >
            questionnaire_question__creation_title
          </div>
          <div
            className="css-12vbzj8-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
            onClick={[MockFunction]}
          />
        </header>
        <section
          className="css-1gs9x9w-Modal"
        >
          <div
            className="css-etbshp-ChooseQuestionnaireQuestionTypeModal"
          >
            <div
              className="css-o2tml0-Card"
              onClick={[Function]}
            >
              <div
                className="card-header css-a0e080-CardHeader"
              >
                <div
                  className="css-1cygpbo-SelectableCard"
                >
                  <div
                    className="css-17tkprm-SelectableCard"
                  >
                    <div
                      className="css-1lrzij5-Icon"
                      dangerouslySetInnerHTML={
                        {
                          "__html": {},
                        }
                      }
                    />
                    <div
                      className="css-1f6p2vj-Heading"
                    >
                      questionnaire_question__multiple_choice_label
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="css-hpovgn-Text"
              >
                questionnaire_question__multiple_choice_description
              </div>
            </div>
            <div
              className="css-o2tml0-Card"
              onClick={[Function]}
            >
              <div
                className="card-header css-a0e080-CardHeader"
              >
                <div
                  className="css-1cygpbo-SelectableCard"
                >
                  <div
                    className="css-17tkprm-SelectableCard"
                  >
                    <div
                      className="css-1lrzij5-Icon"
                      dangerouslySetInnerHTML={
                        {
                          "__html": {},
                        }
                      }
                    />
                    <div
                      className="css-1f6p2vj-Heading"
                    >
                      questionnaire_question__single_choice_label
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="css-hpovgn-Text"
              >
                questionnaire_question__single_choice_description
              </div>
            </div>
            <div
              className="css-o2tml0-Card"
              onClick={[Function]}
            >
              <div
                className="card-header css-a0e080-CardHeader"
              >
                <div
                  className="css-1cygpbo-SelectableCard"
                >
                  <div
                    className="css-17tkprm-SelectableCard"
                  >
                    <div
                      className="css-1lrzij5-Icon"
                      dangerouslySetInnerHTML={
                        {
                          "__html": {},
                        }
                      }
                    />
                    <div
                      className="css-1f6p2vj-Heading"
                    >
                      questionnaire_question__free_text_label
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="css-hpovgn-Text"
              >
                questionnaire_question__free_text_description
              </div>
            </div>
          </div>
        </section>
        <footer
          className="css-ukdqb3-Modal"
        >
          <button
            className="cancel-button css-a1i4d5-Button"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="css-95j2hm-Button"
            >
              cancel_button
            </div>
          </button>
          <button
            className="confirm-button css-1ex7tyh-Button"
            disabled={true}
            onClick={[Function]}
          >
            <div
              className="css-95j2hm-Button"
            >
              confirm_button
            </div>
          </button>
        </footer>
      </div>
    </div>
  </div>
</div>
`;
